File: core/esmf-aspect-meta-model-java/src/main/java/org/eclipse/esmf/characteristic/impl/DefaultSortedSet.java
/*
 * Copyright (c) 2023 Robert Bosch Manufacturing Solutions GmbH
 *
 * See the AUTHORS file(s) distributed with this work for additional
 * information regarding authorship.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package org.eclipse.esmf.characteristic.impl;

import java.util.Optional;
import java.util.StringJoiner;

import org.eclipse.esmf.characteristic.SortedSet;

import org.eclipse.esmf.metamodel.Characteristic;
import org.eclipse.esmf.metamodel.CollectionValue;
import org.eclipse.esmf.metamodel.Type;
import org.eclipse.esmf.metamodel.loader.MetaModelBaseAttributes;
import org.eclipse.esmf.metamodel.visitor.AspectVisitor;

public class DefaultSortedSet extends DefaultCollection implements SortedSet {

   public DefaultSortedSet( final MetaModelBaseAttributes metaModelBaseAttributes,
         final Optional<Type> dataType, final Optional<Characteristic> elementCharacteristic ) {
      super( metaModelBaseAttributes, dataType, false, true, elementCharacteristic );
   }

   @Override
   public CollectionValue.CollectionType getCollectionType() {
      return CollectionValue.CollectionType.SORTEDSET;
   }

   /**
    * Accepts an Aspect visitor
    *
    * @param visitor The visitor to accept
    * @param <T> The result type of the traversal operation
    * @param <C> The context of the visitor traversal
    */
   @Override
   public <T, C> T accept( final AspectVisitor<T, C> visitor, final C context ) {
      return visitor.visitSortedSet( this, context );
   }

   @Override
   public String toString() {
      return new StringJoiner( ", ", DefaultSortedSet.class.getSimpleName() + "[", "]" )
            .toString();
   }
}

File: core/esmf-aspect-model-document-generators/src/test/resources/ts/snapshots/esmf/aspect-meta-model/characteristic/default-sorted-set.ts
/*
 * Copyright (c) 2023 Robert Bosch Manufacturing Solutions GmbH
 *
 * See the AUTHORS file(s) distributed with this work for
 * additional information regarding authorship.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */

import {DefaultCollection} from './default-collection';
import {Characteristic} from './default-characteristic';
import {Type} from '../type';
import { DefaultEntityInstance } from '../default-entity-instance';
import { DefaultEntity } from '../default-entity';
import { Enumeration } from './default-enumeration';

export class DefaultSortedSet extends DefaultCollection implements Enumeration {
    constructor(metaModelVersion: string, aspectModelUrn: string, name: string, elementCharacteristic?: Characteristic, dataType?: Type,private _values?: Array<DefaultEntityInstance | string | number>) {
        super(metaModelVersion, aspectModelUrn, name, false, true, elementCharacteristic, dataType);
    }
    public set values(values: Array<DefaultEntityInstance | string | number>) {
        this._values = values;
    }

    public get values(): Array<DefaultEntityInstance | string | number> {
        return this._values;
    }

        /**
     * Find the index of the given value in the values array.
     *
     * @return returns the index of the value or -1
     */
        public indexOf(value: string): number {
            if (!this.values) {
                return -1;
            }
    
            if (this.dataType && this.dataType.isComplex) {
                const propertyValue = (<DefaultEntity>this.dataType).properties.find(property => property.isNotInPayload === false);
    
                return this.values.findIndex((valueEntry: DefaultEntityInstance) => {
                    return valueEntry[propertyValue.name] === value;
                });
            } else {
                return this.values.findIndex(valueEntry => valueEntry === value);
            }
        }
}
